CONTRIBUTING.md

```diff
@@ -1,7 +1,75 @@
+Some tests contain representative real-world code, however new tests should focus on extensively covering as many edge cases as possible.
+
+Some tests are regression tests, located in `test/correct` and `tests/invalid`.
+The former will simply test that the output never changes, the latter tests parser failures on incorrect input.
+Tests in `test/correct` should be as minimalistic as possible, as to not trip on unrelated changes.
+However this is not always avoidable and they occasionally need manual adjustment.
+
+When editing a test or adding a new one, try to put it into a separate commit as to not mix input changes with diff changes.
+Retrospectively fixing this requires advanced git rebasing skills, and using a helper tool like [lazygit](https://github.com/jesseduffield/lazygit) is strongly recommended.
+
+## Debugging
+
+Short strings can easily be tested with `cabal v2-run --verbose=0 nixfmt -- -w=80 < <(echo $'some code here')`.
+Note the usage of `$''` [ANSI-C quoting](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html#ANSI_002dC-Quoting), which makes it easier to control the line breaks of the input.
+`-w=80` is the default width used for the tests, however sometimes setting it to something ridiculously small can help.
+
+The `--ast` flag outputs a pretty print of the parser result, which is especially helpful when debugging parser bugs.
+
+The `--ir` flag outputs a pretty print of the IR produced by `Pretty.hs`, to inspect it before it gets rendered.
+
+## Architecture overview
+
+The data flow of the formatting process is as follows: (Text) → \[Parser\] → (AST) → \[Pretty\] → (IR) → \[Render\] → (Text).
+
+### AST
+
+The AST is defined in `Types.hs` (entry point type `File`).
+Conceptually it sits somewhere between an annotated AST and a CST: It still uses tokens as primitive, however does not retain enough information to faithfully reproduce the parser input (as would typically be the case for a proper CST).
+
+Every token is annotated with adjacent comments if present, and the line in the source code of the token.
+
+### Pretty
+
+`pretty` transforms each AST element into its IR by recursively walking the tree.
+This is where the actual formatting logic is implemented.
+Effectively, this is a giant switch case on all possible AST values, sometimes taking specific combinations of AST nodes into account for special casing.
+
+### IR and rendering
+
+*If you are familiar with [Wadler/Leijen style pretty printers](https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf) as commonly used in the Haskell ecosystem, this is another instance of them.*
+
+The IR is defined by the `Doc` and `DocE` types in `Predoc.hs`.
+It is a tree structure where the leaves are either text or whitespace and the intermediate nodes group the child nodes together.
+
+The rough idea is that "Spacing" in the IR can be either a whitespace or a line break, and which one will only be decided at rendering time.
```

Comment:
I had to read this a few times. I think I understand it now. I misinterpreted "can be either" as "this is a union of 2 things, but each Spacing is concretely one of these things" , when really Spacing represents a deferred decision (IIUC). Here's a possible way to rephrase this. Feel free workshop further:

```suggestion
The rough idea is that "Spacing" in the IR will be rendered as either a whitespace or a line break. The decision is deferred until rendering time.
```

Reply:
I get your confusion. Does it help if "can be" → "can become"?

Reply:
Yes!